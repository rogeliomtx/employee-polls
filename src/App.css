/* Auth */
body {
  background-color: #f5f5f5;
}

.card.auth {
  width: 18rem;
}

img.avatar, div.avatar {
  width: 100px;
  height: 100px;
  display: block;
  margin: 0 auto;
  border-radius: 100%;
}

div.avatar {
  background: #0d6efd;
  color: #fff;
  border-radius: 100%;
}

div.avatar span {
  font-size: 3em;
  line-height: 100px;
  display: block;
  text-align: center;
}

.navbar .avatar {
  width: 30px;
  height: 30px;
  display: inline-block;
  margin-right: .5em;
}

.leaderboard .avatar {
  width: 50px;
  height: 50px;
  display: inline-block;
  margin-right: .5em;
}

.navbar .logout {
  cursor: pointer;
}