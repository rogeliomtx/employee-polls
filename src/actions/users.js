export const RECEIVE_USERS = 'RECEIVE_USERS';
export const UPDATE_USER_ANSWER = 'UPDATE_USER_ANSWER';
export const ADD_USER_QUESTION = 'ADD_USER_QUESTION';

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function answerUserQuestion(answer) {
  return {
    type: UPDATE_USER_ANSWER,
    answer,
  };
}

export function addUserQuestion(question) {
  return {
    type: ADD_USER_QUESTION,
    question,
  };
}