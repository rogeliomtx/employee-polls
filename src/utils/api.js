import { _getUsers, _getQuestions, _saveQuestion, _saveQuestionAnswer } from './_DATA';

export async function getInitialData () {
  return Promise.all([
    _getUsers(),
    _getQuestions(),
  ]).then(([users, questions]) => ({
    users,
    questions
  }))
}

export async function saveQuestion (question) {
  return _saveQuestion(question)
}

export async function saveQuestionAnswer ({ authedUser, question, answer }) {
  return _saveQuestionAnswer({
    authedUser,
    qid: question.id,
    answer
  }).then((saved) => {
    return { authedUser, answer, questionId: question.id }
  })
}

export async function login (id, password) {
  return _getUsers().then((users) => {
    const user = Object.values(users).find((user) => user.id === id && user.password === password);
    return user;
  })
}
