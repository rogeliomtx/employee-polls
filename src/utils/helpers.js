export function formatQuestion (question, author, authedUser) {
  const { id, optionOne, optionTwo, timestamp } = question

  return {
    id,
    timestamp,
    optionOne,
    optionTwo,
    author,
    hasAnswered: optionOne.votes.includes(authedUser) || optionTwo.votes.includes(authedUser)
  }
}