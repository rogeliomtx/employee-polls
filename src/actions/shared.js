import { showLoading, hideLoading } from 'react-redux-loading-bar';

import { receiveUsers, answerUserQuestion, addUserQuestion } from './users';
import { receiveQuestions, answerQuestion, addQuestion } from './questions';
// import { setAuthedUser } from './authedUser';
import {
  getInitialData as apiGetInitialData,
  saveQuestionAnswer as apiSaveQuestionAnswer,
  saveQuestion as apiSaveQuestion
} from '../utils/api';

// const AUTHED_ID = 'tylermcginnis';

export function handleInitialData() {
  return (dispatch) => {
    dispatch(showLoading());
    return apiGetInitialData()
      .then(({ users, questions }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
        // dispatch(setAuthedUser(AUTHED_ID));
        dispatch(hideLoading());
      });
  }
}

export function handleAnswerQuestion(question, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    showLoading();
    return apiSaveQuestionAnswer({
      authedUser,
      question,
      answer
    })
      .then(({ authedUser, answer, questionId }) => {
        dispatch(answerQuestion({ authedUser, answer, questionId }));
        dispatch(answerUserQuestion({ authedUser, answer, questionId }));
      })
      .then(() => hideLoading());
  }
};

export function handleAddQuestion(optionOne, optionTwo) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    showLoading();
    return apiSaveQuestion({
      optionOneText: optionOne,
      optionTwoText: optionTwo,
      author: authedUser
    })
      .then((question) => {
        dispatch(addQuestion(question));
        dispatch(addUserQuestion(question));
      })
      .then(() => hideLoading());
  };
}