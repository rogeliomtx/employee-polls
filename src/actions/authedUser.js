import { showLoading, hideLoading } from 'react-redux-loading-bar';

import { login as apiLogin } from '../utils/api';

export const SET_AUTHED_USER = 'SET_AUTHED_USER';

export function setAuthedUser(id) {
  return {
    type: SET_AUTHED_USER,
    id,
  };
}

export function handleLogin(id, password) {
  return (dispatch) => {
    showLoading();
    return apiLogin(id, password).then((user) => {
      if (user) {
        dispatch(setAuthedUser(user.id));
        return user;
      }
      return null;
    }).then(() => hideLoading());
  }
}

export function handleLogout() {
  showLoading();
  return (dispatch) => {
    dispatch(setAuthedUser(null));
    hideLoading();
    return true;
  }
}