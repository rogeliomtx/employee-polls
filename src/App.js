import { Route, Routes } from "react-router-dom";
import { useEffect, Fragment } from "react";
import { connect } from "react-redux";
import { LoadingBar } from "react-redux-loading-bar";

import "bootstrap/dist/css/bootstrap.min.css";
import './App.css';

import Home from './components/Home';
import Leaderboard from './components/Leaderboard';
import Login from './components/Login';
import CreateQuestion from './components/CreateQuestion';
import QuestionDetails from './components/QuestionDetails';
import Navbar from './components/Navbar';
import NotFound from './components/NotFound';
import ProtectedRoute from "./components/ProtectedRoute";

import { handleInitialData } from "./actions/shared";

function App({ dispatch, loading }) {
  useEffect(() => {
    dispatch(handleInitialData());
  });

  return (
    <Fragment>
      <LoadingBar />
      {
        loading ? <div>Loading...</div> : (
          <div>
            <Navbar />
            <div className="container">
              <Routes>                
                <Route path="/" element={
                  <ProtectedRoute>
                    <Home />
                  </ProtectedRoute>
                } />
                <Route path="/login" element={<Login />} />
                <Route path="/leaderboard" element={
                  <ProtectedRoute>
                    <Leaderboard />
                  </ProtectedRoute>
                } />
                <Route path="/add" element={
                  <ProtectedRoute>
                    <CreateQuestion />
                  </ProtectedRoute>
                } />
                <Route path="/questions/:questionId" element={
                  <ProtectedRoute>
                    <QuestionDetails />
                  </ProtectedRoute>
                } />
                <Route path="/*" element={<NotFound />} />
              </Routes>
            </div>
          </div>
        )
      }
    </Fragment>
  );
}

const mapStateToProps = ({ users, authedUser }) => {
  return {
    loading: users === null,  // if users are loaded, loading is false
    authedUser,
  }
}
export default connect(mapStateToProps)(App);
