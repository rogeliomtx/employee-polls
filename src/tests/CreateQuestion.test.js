import { MemoryRouter } from "react-router";
import { screen, waitFor, fireEvent, within, act } from '@testing-library/react';

import { renderWithProviders, getMockState } from "./utils";
import App from "../App";
import { handleAddQuestion } from '../actions/shared';

const authedUserState = {preloadedState: getMockState("sarahedo")};

const component = (
  <MemoryRouter initialEntries={["/add"]}>
    <App />
  </MemoryRouter>
);

describe("CreateQuestion", () => {
  it("should match the snapshot", () => {
    const view = renderWithProviders(component, authedUserState);
    expect(view.container).toMatchSnapshot();
  });

  it ("should render the CreateQuestion form", () => {
    renderWithProviders(component, authedUserState);

    expect(screen.getByText(/would you rather/i)).toBeInTheDocument();
    expect(screen.getByText(/option 1/i)).toBeInTheDocument();
    expect(screen.getByText(/option 2/i)).toBeInTheDocument();
    expect(screen.getByRole('button', {name: /submit/i})).toBeInTheDocument();

    expect(screen.getAllByRole('textbox').length).toBe(2);
  });

  // test question creation
  it ("should create a new question", async () => {
    renderWithProviders(component, authedUserState);

    const optionOne = screen.getAllByRole('textbox')[0];
    const optionTwo = screen.getAllByRole('textbox')[1];
    const submitBtn = screen.getByRole('button', {name: /submit/i});

    fireEvent.change(optionOne, {target: {value: "Option One"}});
    fireEvent.change(optionTwo, {target: {value: "Option Two"}});

    fireEvent.click(submitBtn);

    await waitFor(() => {
      const unansweredQuestions = screen.getByTestId("unanswered-questions");
      const questions = within(unansweredQuestions).queryAllByRole("article");
      expect(questions.length).toBe(3);
    });
  });
});

describe("CreateQuestion fns", () => {
  it("should create a new question", () => {
    const { store } = renderWithProviders(component, authedUserState);
    const optionOne = "Option One";
    const optionTwo = "Option Two"

    act(() => {
      store.dispatch(handleAddQuestion(optionOne, optionTwo)).then(() => {
        const { questions } = store.getState();
        
        // check questions
        const question = questions[Object.keys(questions)[0]];
        expect(question.optionOne.text).toBe(optionOne);
        expect(question.optionTwo.text).toBe(optionTwo);
      });
    });

    
  });
});