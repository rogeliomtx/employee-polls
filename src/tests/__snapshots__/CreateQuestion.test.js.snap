// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateQuestion should match the snapshot 1`] = `
<div>
  <div />
  <div>
    <nav
      class="navbar navbar-expand-lg bg-primary"
      data-testid="menu"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand link-light"
          href="/"
        >
          Employee Polls
        </a>
        <div
          class="collapse navbar-collapse"
          id="navbarText"
        >
          <ul
            class="navbar-nav me-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link active link-light"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-light"
                href="/add"
              >
                Add question
              </a>
            </li>
          </ul>
          <ul
            class="navbar-nav mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <span
                class="nav-link link-light"
              >
                <div
                  class="avatar"
                >
                  <img
                    alt="Avatar"
                    class="avatar"
                    src="/assets/avatar_1.jpg"
                  />
                </div>
                Sarah Edo
              </span>
            </li>
            <li
              class="nav-item"
            >
              <span
                class="nav-link link-light logout"
              >
                Logout
              </span>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        class="card mt-5"
        role="contentinfo"
      >
        <div
          class="card-header"
        >
          Add Your Own Poll
        </div>
        <div
          class="card-body"
        >
          <h1>
            Would you rather...
          </h1>
          <form>
            <div
              class="form-group mt-3"
            >
              <label
                class="form-label"
              >
                Option 1
              </label>
              <input
                class="form-control"
                name="optionOne"
                placeholder="Enter option one text here"
                type="text"
              />
            </div>
            <div
              class="form-group mt-3"
            >
              <label
                class="form-label"
              >
                Option 2
              </label>
              <input
                class="form-control"
                name="optionTwo"
                placeholder="Enter option two text here"
                type="text"
              />
            </div>
            <div
              class="mt-3"
            >
              <button
                class="btn btn-primary btn-lg"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
